src/Core/Settings.cpp

```diff
@@ -1346,6 +1346,17 @@ Possible values:
 
 - 0 — Disabled.
 - 1 — Enabled.
+)", 0) \
+    M(UInt64, skip_indexes_in_final_correctness_threshold, 0, R"(
+Controls whether granules returned by a skipping index are expanded to return correct results when executing a query with the FINAL modifier.
+
+Using skip indexes may exclude rows (granules) containing the latest data which could lead to incorrect results. This setting can ensure that correct results are returned by scanning newer parts that have overlap with the ranges returned by the skip index. (Experimental)
+
+Possible values:
+
+- 0 — Disabled.
+- 1 — Enabled.
+>= 2 - Limit on the number of granules returned by the skip index, beyond which feature is disabled.
```

Comment:
Seems a bit unusual, is there any other setting that have this multiple special values together with simple semantics of numerical value?

Maybe reconsider this one?

Reply:
Sure, good point, I just went for a quick fix solution.

A knob is needed to turn ON/OFF the fix. It is perfectly safe to set the existing `use_skip_indexes_if_final `to 1 if it is known that the secondary index column value **never mutates** after insert (even if other column values are updated). Thus enabling this new fix by default will cause a performance regression as more rows could be scanned than current.

Some additional protection maybe required if the first step secondary index lookup returns 100's of granules. The `KeyCondition `for the 2nd step could then contain 100s of RPN nodes and take memory / CPU for the repeated searches. Hence some setting for _exiting_ out of the fix is needed if there are too many ranges returned by the first step. An example of the KeyCondition where the primary key has 3 column and 12 granules are returned in the first pass :-

```
(column 0 in ['303ed4c698', '32b30a250a']), (column 1 in [80, 99]), and, (column 2 in [647080, 809299]), and,
(column 0 in ['5ea1649a31', '5fd0b37cd7']), (column 1 in [64, 83]), and, (column 2 in [537064, 114283]), and, 
or, (column 0 in ['6da37dd313', '6f2268bd1d']), (column 1 in [17, 36]), and, (column 2 in [331617, 753836]), and,
or, (column 0 in ['7143d7fbad', '7380ad8a67']), (column 1 in [56, 75]), and, (column 2 in [781056, 497275]), and,
or, (column 0 in ['8f468c873a', '912d2b1c7b']), (column 1 in [5, 24]), and, (column 2 in [968605, 751824]), and, 
or, (column 0 in ['c15da1f2b5', 'c32d9bf27a']), (column 1 in [47, 66]), and, (column 2 in [788247, 979466]), and, 
or, (column 0 in ['5b8add2a5d', '5ea1649a31']), (column 1 in [44, 64]), and, (column 2 in [317844, 537064]), and, 
or, (column 0 in ['a1d0c6e83f', 'a4a042cf4f']), (column 1 in [78, 97]), and, (column 2 in [42478, 171697]), and, 
or, (column 0 in ['a64c94baaf', 'a87ff679a2']), (column 1 in [16, 36]), and, (column 2 in [898816, 4036]), and, 
or, (column 0 in ['05049e90fa', '0640966322']), (column 1 in [57, 76]), and, (column 2 in [367657, 158876]), and, 
or, (column 0 in ['7b13b22030', '7eabe3a164']), (column 1 in [13, 32]), and, (column 2 in [993613, 206832]), and, 
or, (column 0 in ['f4be00279e', 'f7e9050c92']), (column 1 in [27, 65]), and, (column 2 in [528227, 854665]), and, or
```
I will rework this setting stuff.